### DISCUSSION TITLE
Deployment error: Cannot read properties of undefined (reading 'currentPicture') but working on localhost

### DISCUSSION THREAD

**Discussion by Elgin-Schultz** [neuefische/questions#457] [Web] [answered]
I get following error message, but on local host it is working. What am I doing wrong? Thank you for your help.
<img width="1223" alt="Bildschirm­foto 2023-01-19 um 12 15 52" src="https://user-images.githubusercontent.com/115506885/213429560-ba617bf0-6511-467e-814a-c4b2cf9fcff8.png">

Here is my [PR](https://github.com/Elgin-Schultz/capstone-project/pull/21)

**Comment by julialoeschel** [comment]
please share the PR, so that we can have a look at your code 👍

**Reply by Elgin-Schultz** [reply]
https://github.com/Elgin-Schultz/capstone-project/pull/21

**Comment by andy-schwedler** [comment]
Hey Elgin, wenn ich das richtig sehe, dann hat er hier ein Problem mit `currentPicture` - das ist `undefined`.

Lass es dir mal in der Console ausloggen, vielleicht kannst du es dann leichter zurückverfolgen?

**Comment by ufelgen** [comment]
Hi Elgin, versuch mal currentPicture?.xxx mit dem Fragezeichen oder alternativ if (!currentPicture) {return null}

**Comment by Elgin-Schultz** [comment]
}
  if (!currentSite.currentPicture) {
    return null;
  }

  if (!siteCurrents) {
    return null;
  }

I put these snippets in, but unfortunately it didn't solve the problem...

**Reply by ufelgen** [reply]
in Zeile 85 vielleicht `{siteCurrents?.map((current) =>` ??

**Reply by Elgin-Schultz** [reply]
This solved the problem:

import { allSites } from "../../lib/data-sites";
export default function singleSite() {
  const router = useRouter();
  const slug = router.query.slug;
  if (!slug) {
    return null;
  }
  const currentSite = allSites.find((site) => site.slug === slug);


Thank you Thomas &Julia!!!!!

**Comment by Elgin-Schultz** [comment, marked as answer]
This solved the problem:

import { allSites } from "../../lib/data-sites";
export default function singleSite() {
const router = useRouter();
const slug = router.query.slug;
if (!slug) {
return null;
}
const currentSite = allSites.find((site) => site.slug === slug);

Thank you Thomas &Julia!!!!!

**Comment by mpagels** [comment]
The problem is, how you want to use the `[slug].js` page in `pages/singlesite` and how `Next` is working under the hood.

At `build` time, when `vercel` is deploying your project, [slug].js is empty, because no one is visiting the site and next doesn't know what `[slug]` holds. And therefor the build process is throwing the error.

You can tell Next how to handle `dynamic path` during the `build` process. Like -> "hey next, I know every `slug/id` and have every information here to build `singlesite/1` to `singlesite/7`.  Or you don't know it and next needs to build the pages at every request.

And `Next` has you covered. Take a look at [getStaticPaths](https://nextjs.org/docs/basic-features/data-fetching/get-static-paths) from the official next documentation.

In your example, it should like this

(I removed the `useRouter` completely from the file)

```js
export default function singleSite({ slug }) {
  const currentSite = allSites.find((site) => site.slug === slug);
  const siteCurrents = currentSite.currentPicture;
  const ancientPictArr = currentSite.ancientPicture;
  const siteInformations = currentSite.information;
  const siteCredits = currentSite.credits;

// ... your other code

}
// after the function and before the styled components part

// Generates `/singlesite/1` to `/singlesite/7`
export async function getStaticPaths() {
  return {
    paths: [
      { params: { slug: "1" } },
      { params: { slug: "2" } },
      { params: { slug: "3" } },
      { params: { slug: "4" } },
      { params: { slug: "5" } },
      { params: { slug: "6" } },
      { params: { slug: "7" } },
    ],
    fallback: false, // can also be true or 'blocking'
  };
}

// `getStaticPaths` requires using `getStaticProps`
export async function getStaticProps(context) {
  return {
    // Passed to the page component as props
    props: { slug: context.params.slug },
  };
}

// here begins the styled component part. That is not important, just for you to know where I am in your code. You could put the whole getStaticPaths/getStaticProps thing on the very end after the styled component and it should work.
```

You could also do the "extracting and finding" logic part here in the `getStaticProps` function and pass the information as props. Makes the  function way cleaner.

So now you have a page, that is completely built at `build` time and super fast for your end user (like doesn't need to be build at runtime, when the user is visiting them)

If you want learn more how next work on this, read the documentation completely from [Pages](https://nextjs.org/docs/basic-features/pages) to `Data Fetching`.

**Reply by Elgin-Schultz** [reply]
Thank you so much!!!!!
